### PR TITLE
Add fallback when both co- and contra-variant inference candidates exist

### LINKED ISSUE #54005: Inference regression (5.0.4 vs 4.9.5) with nullable enum function parameter
# Bug Report

### 🔎 Search Terms

inference string null function parameter 

### 🕗 Version & Regression Information

- This changed between versions 4.9.5 and 5.0.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/KYOwrgtgBAKgFgSxAcygbygQygXigck3wBooAjXAskqAXwG4AoRgMzBAGMAXBAexCgsAFIyhioAJ2DIEAZy4TMPfgGUuSsLIDyAB2UhZALigrgAG2Ddd+2QB54SZAD5io8fwDCcTCmDGh8hpGsIgoUAA+UOBmZgCUuE5QAG68CAAmpAD0mVC2ALR5UABKwBC8ScBQXHCVAAaR0Wa1WCBpVTVQwBISvBJQyLzAslgA7pgAngB0jLHGKenobmKy5pZcQhjZuQVQAIJpbT6dAB46ZggcCFxV4zqVmBLIkKDXtQ4ozT5t1ZVdPX0DIajCbTcRgqC8PR8AzGKQyeSKfRqILWaGyVzg9wgLw+ZB+CHY7y+LI5ACiRSKWiKUCQyQArJMAAyTAAs5DA1xAvGutKSLMmAE5JnTjDBbpV8AF1FxNKLQqgGmAYvEcIl5ml8DThlzrphZLIEMgQJgyBYqrwbncCEIANbAcbGBGOFVq1Ia0HiWixJi0ZjAU69a5sTj6KArCzcWwAaXtMBOXFAaWGTpQTiEOh6OmCpgjXAACpm7DHxjAnLNkm7FmDqj0RlFgHXSd1ekJ8DUYrx8N7GL7GP6dIHLZUc2sC5Ci7H44nkwpHIk8GglgScb5-HaHVBizAXRX0kwwZCbAB+YwjqxQ-gTktOehQLb5QolMoVdp1I+fVqvzrNgGDYaYMYph7Jg+wDCRri4cUTFWc8bGjSd-QTVoZwkOdcCXLYPF4CAIBeTorhqPpeBYL8yHMXg63YaEoFwiAyIkYZeA5Fpvg6P5en6P9gSAsFIl2boJlsNB12MLdaCcJc+IE8Z4OvJggA)

### 💻 Code

```ts
enum Thing { a = 'a', b = 'b', };

function f(
    registrationStatusOptions: SelectOptions<Thing>,
    onChange: (status: Thing | null) => void, // <-- Remove the `| null` and the error goes away.
): void {
    select({ // <-- Add an explicit type argument `Thing` and the error goes away.
        options: registrationStatusOptions,
        onChange: onChange, // ERROR in v5.0.4 but not in v4.9.5: Type '(status: Thing | null) => void' is not assignable to type '(key: string) => void'.
    });
}

export function select<KeyT extends string>(props: SelectProps<KeyT>): void {
    throw new Error('hello');
}

export type SelectProps<KeyT extends string> = {
    onChange: (key: KeyT) => void;
    options?: SelectOptions<KeyT>; // <-- Remove the `?` and the error goes away.
};

export type SelectOptions<KeyT extends string> =
    // Comment either of the below union members out and the error goes away.
    | Array<{key: KeyT}>
    | Array<KeyT>;
```

### 🙁 Actual behavior

1. Got a type error; see "ERROR" comment above. Looks like the type argument to `select(...)` is being inferred incorrectly.
2. Minor perturbations of the code cause the error to go away; see other comments above.

This was extracted from some React UI code.

### 🙂 Expected behavior

TypeScript 4.9.5 infers the type argument I wanted. I was hoping TypeScript 5.0.4 would do the same.

### PR DESCRIPTION
**Pull request by ahejlsberg**
In cases where both co- and contra-variant inference candidates exist, this PR adds the ability to fall back to the secondary inference when the constraint check fails for the primary inference.

Fixes #54005.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by jakebailey** [comment]
@typescript-bot pack this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -25154,22 +25154,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
                         some(inference.contraCandidates, t => isTypeSubtypeOf(inferredCovariantType, t)) &&
-                        every(context.inferences, other =>
-                            other !== inference && getConstraintOfTypeParameter(other.typeParameter) !== inference.typeParameter ||
-                            every(other.candidates, t => isTypeSubtypeOf(t, inferredCovariantType)));
-                    inferredType = useCovariantType ? inferredCovariantType : getContravariantInference(inference);
-                }
-                else if (inferredCovariantType) {
-                    inferredType = inferredCovariantType;
+                        every(context.inferences, other => other === inference ||
+                            getConstraintOfTypeParameter(other.typeParameter) !== inference.typeParameter ||
```
Comment (jakebailey) [inline src/compiler/checker.ts:25171]: Part of #52180 was motivated by noticing that when we treat the "current" inference as special, we incur some order-dependent behavior. Where this came up (if my memory was correct) was when I was editing the `strictFunctionTypes` PR, where I could edit the file and see errors change around non-deterministically.

I tested restoring the change from #52180 and this PR still passes (i.e. putting back `other !== inference && getConstraintOfTypeParameter(other.typeParameter) !== inference.typeParameter`), which makes me feel like reverting it back to the state in #52123 that skips over `other === inference` will bring back that oddity (and we just don't have a test for it).

I'll try and see if I can make it break again like I did the first time.

**Review by jakebailey — APPROVED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:25171]: Looking carefully at the code you had in #52180 and the original code plus the line I added [here](https://github.com/microsoft/TypeScript/pull/54072/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R25168), I've convinced myself they're just two ways of saying the same thing. So I'll restore the change from #52180.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

File: src/compiler/checker.ts (hunk 1 above)
Comment (jakebailey) [reply src/compiler/checker.ts]: Oh! You're right, I missed that new line.

Comment (jakebailey) [reply src/compiler/checker.ts]: In that case, I'm happy with whichever is more performant or clearer. They're definitely equivalent now that I think closely.

**Pull request merged**